<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>Interactive analytics and data visualisation in R</title>
    <link rel="stylesheet" href="../../libraries/frameworks/minimal/stylesheets/styles.css">
    <link rel="stylesheet" href="../../libraries/highlighters/highlight.js/css/solarized_light.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <link rel="stylesheet" href="http://maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css">
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
        
    
  </head>
  <body>
    <div class="wrapper">
<header>
  <h1>Interactive analytics and data visualisation in R</h1>
<!--  <p></p>

   <img width='150px' src='../../assets/img/'></img>
      -->
  <p class="view">
    <!-- <a href="http://github.com/garthtarr/visR">
     View the Project on GitHub <small>garthtarr/visR</small>
    </a>
    <img width='150px' src='../../assets/img/'></img>
    -->
    <a href="http://garthtarr.com">Garth Tarr</a> <span style="float:right;"><a href="https://twitter.com/garthtarr"><i class="fa fa-twitter"></i></a> <a href="https://www.linkedin.com/in/garthtarr"><i class="fa fa-linkedin-square"></i></a> <a href="https://github.com/garthtarr"><i class="fa fa-github"></i></a></span>
  </p>
  
  
      <!--
  <ul class='buttons'>
  <i class="fa fa-twitter"></i><strong><a href="https://twitter.com/garthtarr"> @garthtarr </a></strong>
  </ul>

    <li><a href="https://github.com/garthtarr/visR/zipball/gh-pages">
      Download <strong>ZIP File</strong>
    </a></li>

    <li><a href="https://github.com/garthtarr/visR/tarball/gh-pages">
      View On <strong>GitHub</strong>
    </a></li>

    <li><a href="http://github.com/garthtarr/visR">
      Fork On <strong>GitHub</strong></a>
    </li>
  </ul>
      -->
    <ol>
    </ol>
</header>

      <section>
        <h2>Case study: gene expression in starvation</h2>
<p>Through the process of gene regulation, a cell can control which genes are transcribed from DNA to RNA- what we call being &quot;expressed&quot;. (If a gene is never turned into RNA, it may as well not be there at all). This provides a sort of &quot;cellular switchboard&quot; that can activate some systems and deactivate others, which can speed up or slow down growth, switch what nutrients are transported into or out of the cell, and respond to other stimuli. A gene expression microarray lets us measure how much of each gene is expressed in a particular condition. We can use this to figure out the function of a specific gene (based on when it turns on and off), or to get an overall picture of the cell&#39;s activity.</p>

<p>Brauer 2008 used microarrays to test the effect of starvation and growth rate on bakerâ€™s yeast (S. cerevisiae, a popular model organism for studying molecular genomics because of its simplicity). Basically, if you give yeast plenty of nutrients (a rich media), except that you sharply restrict its supply of one nutrient, you can control the growth rate to whatever level you desire (we do this with a tool called a chemostat). For example, you could limit the yeast&#39;s supply of glucose (sugar, which the cell metabolizes to get energy and carbon), of leucine (an essential amino acid), or of ammonium (a source of nitrogen).</p>

<p>&quot;Starving&quot; the yeast of these nutrients lets us find genes that:</p>

<ul>
<li><strong>Raise or lower their activity in response to growth rate.</strong> Growth-rate dependent expression patterns can tell us a lot about cell cycle control, and how the cell responds to stress.</li>
<li><strong>Respond differently when different nutrients are being limited.</strong> These genes may be involved in the transport or metabolism of those nutrients.</li>
</ul>

<h3>Featured R packages</h3>

<pre><code class="r">install.packages(c(&quot;readr&quot;, &quot;tidyr&quot;, &quot;dplyr&quot;))
</code></pre>

<h3>Get the data and tidy it</h3>

<pre><code class="r">require(readr)
</code></pre>

<pre><code>## Loading required package: readr
</code></pre>

<pre><code class="r">original_data = read_delim(&quot;~/Downloads/Brauer2008_DataSet1.tds&quot;, delim = &quot;\t&quot;)
</code></pre>

<pre><code>## Error: &#39;~/Downloads/Brauer2008_DataSet1.tds&#39; does not exist.
</code></pre>

<pre><code class="r">## View(original_data)  # opens a spreadsheet view in RStudio
dim(original_data)
</code></pre>

<pre><code>## Error in eval(expr, envir, enclos): object &#39;original_data&#39; not found
</code></pre>

<p>Fix the name column by splitting on <code>||</code>, remove white space and drop unecessary variables.  We also want to ensure that we have <a href="http://vita.had.co.nz/papers/tidy-data.pdf">tidy data</a> -- each variable should be one column - the column headers are values not variable names.</p>

<pre><code class="r">require(dplyr)
</code></pre>

<pre><code>## Loading required package: dplyr
## 
## Attaching package: &#39;dplyr&#39;
## 
## The following objects are masked from &#39;package:stats&#39;:
## 
##     filter, lag
## 
## The following objects are masked from &#39;package:base&#39;:
## 
##     intersect, setdiff, setequal, union
</code></pre>

<pre><code class="r">require(tidyr)
</code></pre>

<pre><code>## Loading required package: tidyr
</code></pre>

<pre><code class="r">cleaned_data = original_data %&gt;% separate(NAME, c(&quot;name&quot;, &quot;BP&quot;, &quot;MF&quot;, &quot;systematic_name&quot;, 
    &quot;number&quot;), sep = &quot;\\|\\|&quot;) %&gt;% mutate_each(funs(trimws), name:systematic_name) %&gt;% 
    select(-number, -GID, -YORF, -GWEIGHT) %&gt;% gather(sample, expression, G0.05:U0.3) %&gt;% 
    separate(sample, c(&quot;nutrient&quot;, &quot;rate&quot;), sep = 1, convert = TRUE)
</code></pre>

<pre><code>## Error in eval(expr, envir, enclos): object &#39;original_data&#39; not found
</code></pre>

<p>The above code chunk is doing a lot of processing very sucinctly using the pipe operator (see the <strong>magrittr</strong> package for details).  The <code>gather()</code> function <strong>melts</strong> the data - instead of one row per gene, we now have one row per gene per sample.  We&#39;ve <strong>gathered</strong> 36 columns together into two variables.</p>

<h3>Visualise the data</h3>

<h4>Classical approach</h4>

<pre><code class="r">library(ggplot2)
cleaned_data %&gt;% filter(name == &quot;LEU1&quot;) %&gt;% ggplot(aes(rate, expression, color = nutrient)) + 
    geom_line()
</code></pre>

<pre><code>## Error in eval(expr, envir, enclos): object &#39;cleaned_data&#39; not found
</code></pre>

<pre><code class="r">cleaned_data %&gt;% filter(BP == &quot;leucine biosynthesis&quot;) %&gt;% ggplot(aes(rate, expression, 
    color = nutrient)) + geom_line() + facet_wrap(~name)
</code></pre>

<pre><code>## Error in eval(expr, envir, enclos): object &#39;cleaned_data&#39; not found
</code></pre>

<pre><code class="r">cleaned_data %&gt;% filter(BP == &quot;leucine biosynthesis&quot;) %&gt;% ggplot(aes(rate, expression, 
    color = nutrient)) + geom_point() + geom_smooth(method = &quot;lm&quot;, se = FALSE) + 
    facet_wrap(~name)
</code></pre>

<pre><code>## Error in eval(expr, envir, enclos): object &#39;cleaned_data&#39; not found
</code></pre>

<pre><code class="r">cleaned_data %&gt;% filter(BP == &quot;sulfur metabolism&quot;) %&gt;% ggplot(aes(rate, expression, 
    color = nutrient)) + geom_point() + geom_smooth(method = &quot;lm&quot;, se = FALSE) + 
    facet_wrap(~name + systematic_name, scales = &quot;free_y&quot;)
</code></pre>

<pre><code>## Error in eval(expr, envir, enclos): object &#39;cleaned_data&#39; not found
</code></pre>

<h4>Interactive approach: SHINY!!</h4>

<p>Convert this basic shiny app from <code>ggplot2</code> to <code>ggvis</code>.  The aim here is two fold: 
1. understand how the components link within the <code>shiny</code> app; and 
2. Move beyond static plots with <code>ggvis</code>.</p>

<p>Try adding a line of best fit using <code>geom_smooth()</code> in <code>ggplot2</code> or <code>layer_model_predictions()</code> in <code>ggvis</code>.</p>

<h2>Reference</h2>

<ul>
<li>Robinson D (2015). &quot;Cleaning and visualizing genomic data: a case study in tidy analysis&quot;, blog post. <a href="http://varianceexplained.org/r/tidy-genomics/">http://varianceexplained.org/r/tidy-genomics/</a></li>
<li>Brauer et al. (2008). Coordination of Growth Rate, Cell Cycle, Stress Response, and Metabolic Activity in Yeast,</li>
</ul>

      </section>
      <footer>
        <p>Maintained by <a href="http://github.com/garthtarr"><i class="fa fa-github"></i>/garthtarr
        </a></p>
        <!--<p><small>Hosted on GitHub Pages &mdash; Theme by 
          <a href="https://github.com/orderedlist">orderedlist</a></small>
        </p>-->
      </footer>    </div>
    <script src="../../libraries/frameworks/minimal/javascripts/scale.fix.js"></script>
  </body>
  <!-- Load Javascripts for Widgets -->
  
  <!-- MathJax: Fall back to local if CDN offline but local image fonts are not supported (saves >100MB) -->
  <script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      tex2jax: {
        inlineMath: [['$','$'], ['\\(','\\)']],
        processEscapes: true
      }
    });
  </script>
  <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/2.0-latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
  <!-- <script src="https://c328740.ssl.cf1.rackcdn.com/mathjax/2.0-latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
  </script> -->
  <script>window.MathJax || document.write('<script type="text/x-mathjax-config">MathJax.Hub.Config({"HTML-CSS":{imageFont:null}});<\/script><script src="../../libraries/widgets/mathjax/MathJax.js?config=TeX-AMS-MML_HTMLorMML"><\/script>')
</script>
<!-- LOAD HIGHLIGHTER JS FILES -->
  <script src="../../libraries/highlighters/highlight.js/highlight.pack.js"></script>
  <script>hljs.initHighlightingOnLoad();</script>
  <!-- DONE LOADING HIGHLIGHTER JS FILES -->
   
  </html>