<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>Data visualisation, interactive data analysis, statistical programming</title>
    <link rel="stylesheet" href="../../libraries/frameworks/minimal/stylesheets/styles.css">
    <link rel="stylesheet" href="../../libraries/highlighters/highlight.js/css/solarized_light.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <link rel="stylesheet" href="http://maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css">
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
        
    
  </head>
  <body>
    <div class="wrapper">
<header>
  <h1>Data visualisation, interactive data analysis, statistical programming</h1>
<!--  <p></p>

   <img width='150px' src='../../assets/img/'></img>
      -->
  <p class="view">
    <!-- <a href="http://github.com/garthtarr/visR">
     View the Project on GitHub <small>garthtarr/visR</small>
    </a>
    <img width='150px' src='../../assets/img/'></img>
    -->
    <a href="http://garthtarr.com">Garth Tarr</a> <span style="float:right;"><a href="https://twitter.com/garthtarr"><i class="fa fa-twitter"></i></a> <a href="https://www.linkedin.com/in/garthtarr"><i class="fa fa-linkedin-square"></i></a> <a href="https://github.com/garthtarr"><i class="fa fa-github"></i></a></span>
  </p>
  
  
      <!--
  <ul class='buttons'>
  <i class="fa fa-twitter"></i><strong><a href="https://twitter.com/garthtarr"> @garthtarr </a></strong>
  </ul>

    <li><a href="https://github.com/garthtarr/visR/zipball/gh-pages">
      Download <strong>ZIP File</strong>
    </a></li>

    <li><a href="https://github.com/garthtarr/visR/tarball/gh-pages">
      View On <strong>GitHub</strong>
    </a></li>

    <li><a href="http://github.com/garthtarr/visR">
      Fork On <strong>GitHub</strong></a>
    </li>
  </ul>
      -->
    <ol>
    </ol>
</header>

      <section>
        <h2>Case study: gene expression in starvation</h2>
<p>Through the process of gene regulation, a cell can control which genes are transcribed from DNA to RNA- what we call being &quot;expressed&quot;. (If a gene is never turned into RNA, it may as well not be there at all). This provides a sort of &quot;cellular switchboard&quot; that can activate some systems and deactivate others, which can speed up or slow down growth, switch what nutrients are transported into or out of the cell, and respond to other stimuli. A gene expression microarray lets us measure how much of each gene is expressed in a particular condition. We can use this to figure out the function of a specific gene (based on when it turns on and off), or to get an overall picture of the cell&#39;s activity.</p>

<p>Brauer (2008) used microarrays to test the effect of starvation and growth rate on baker’s yeast (S. cerevisiae, a popular model organism for studying molecular genomics because of its simplicity). Basically, if you give yeast plenty of nutrients (a rich media), except that you sharply restrict its supply of one nutrient, you can control the growth rate to whatever level you desire (we do this with a tool called a chemostat). For example, you could limit the yeast&#39;s supply of glucose (sugar, which the cell metabolizes to get energy and carbon), of leucine (an essential amino acid), or of ammonium (a source of nitrogen).</p>

<p>&quot;Starving&quot; the yeast of these nutrients lets us find genes that:</p>

<ul>
<li><strong>Raise or lower their activity in response to growth rate.</strong> Growth-rate dependent expression patterns can tell us a lot about cell cycle control, and how the cell responds to stress.</li>
<li><strong>Respond differently when different nutrients are being limited.</strong> These genes may be involved in the transport or metabolism of those nutrients.</li>
</ul>

<h3>Featured R packages</h3>

<pre><code class="r">install.packages(c(&quot;readr&quot;, &quot;tidyr&quot;, &quot;dplyr&quot;))
</code></pre>

<h3>Get the data and tidy it</h3>

<p>The data, tidying code and examples are borrowed from <a href="http://varianceexplained.org/r/tidy-genomics/">here</a> and <a href="http://varianceexplained.org/r/tidy-genomics-broom/">here</a>.</p>

<pre><code class="r">require(readr)
original_data = read_delim(&quot;http://www.maths.usyd.edu.au/u/gartht/Brauer2008_DataSet1.tds&quot;, 
    delim = &quot;\t&quot;)
## View(original_data)  # opens a spreadsheet view in RStudio
dim(original_data)
</code></pre>

<pre><code>## [1] 5537   40
</code></pre>

<p>Fix the name column by splitting on <code>||</code>, remove white space and drop unecessary variables.  We also want to ensure that we have <a href="http://vita.had.co.nz/papers/tidy-data.pdf">tidy data</a> -- each variable should be one column - the column headers are values not variable names.</p>

<pre><code class="r">require(dplyr)
require(tidyr)

nutrient_names &lt;- c(G = &quot;Glucose&quot;, L = &quot;Leucine&quot;, P = &quot;Phosphate&quot;,
               S = &quot;Sulfate&quot;, N = &quot;Ammonia&quot;, U = &quot;Uracil&quot;)


cleaned_data = original_data %&gt;%
  separate(NAME, 
           c(&quot;name&quot;, &quot;BP&quot;, &quot;MF&quot;, &quot;systematic_name&quot;, &quot;number&quot;), 
           sep = &quot;\\|\\|&quot;) %&gt;%
  mutate_each(funs(trimws), name:systematic_name) %&gt;%
  select(-number, -GID, -YORF, -GWEIGHT)  %&gt;%
  gather(sample, expression, G0.05:U0.3) %&gt;%
  separate(sample, c(&quot;nutrient&quot;, &quot;rate&quot;), sep = 1, convert = TRUE) %&gt;%
  mutate(nutrient = plyr::revalue(nutrient, nutrient_names)) %&gt;%
  filter(!is.na(expression), systematic_name != &quot;&quot;)
names(cleaned_data)
</code></pre>

<pre><code>## [1] &quot;name&quot;            &quot;BP&quot;              &quot;MF&quot;              &quot;systematic_name&quot;
## [5] &quot;nutrient&quot;        &quot;rate&quot;            &quot;expression&quot;
</code></pre>

<p>The above code chunk is doing a lot of processing very sucinctly using the pipe operator (see the <a href="https://cran.r-project.org/web/packages/magrittr/vignettes/magrittr.html"><code>magrittr</code></a> package for details).  The <code>gather()</code> function <strong>melts</strong> the data - instead of one row per gene, we now have one row per gene per <code>sample</code>.  We&#39;ve <strong>gathered</strong> 36 columns together into two variables (<code>expression</code> and <code>sample</code>) then <strong>separated</strong> <code>sample</code> out into two variables (<code>nutrient</code> and <code>rate</code>).  Note that <code>sample</code> never appears in the final output... the wonders of the pipe (<code>%&gt;%</code>) operator.</p>

<h3>Visualise the data</h3>

<p>Below is a classical approach using <code>ggplot2</code>.</p>

<p>Look at one gene <strong>LEU1</strong>:</p>

<pre><code class="r">library(ggplot2)
cleaned_data %&gt;%
  filter(name == &quot;LEU1&quot;) %&gt;%
  ggplot(aes(rate, expression, color = nutrient)) +
  geom_line()
</code></pre>

<p><img src="assets/fig/unnamed-chunk-4-1.png" alt="plot of chunk unnamed-chunk-4"> </p>

<p><strong>LEU1</strong>&#39;s expression is far higher (more “turned on”) when the cell is being starved of leucine than in any other condition, because in that case the cell has to synthesize its own leucine. And as the amount of leucine in the environment (the growth rate) increases, the cell can focus less on leucine production, and the expression of those genes go down. We’ve just gotten one snapshot of our gene’s regulatory network, and how it responds to external stimuli.</p>

<p>To look at all the genes in the <strong>leucine biosynthesis</strong> process subset using the <code>BP</code> variable, to filter for all genes in that process, and then facet to create sub-plots for each.</p>

<pre><code class="r">cleaned_data %&gt;%
  filter(BP == &quot;leucine biosynthesis&quot;) %&gt;%
  ggplot(aes(rate, expression, color = nutrient)) +
  geom_line() +
  facet_wrap(~name)
</code></pre>

<p><img src="assets/fig/unnamed-chunk-5-1.png" alt="plot of chunk unnamed-chunk-5"> </p>

<p><strong>LEU1</strong>, <strong>LEU2</strong>, and <strong>LEU4</strong> all show a similar pattern, where starvation of leucine causes higher gene expression. <strong>LEU4</strong> also appears to respond to glucose starvation as well. </p>

<p>We can do some basic model fitting, adding lines of best fit using  <code>geom_smooth(method = &quot;lm&quot;, se = FALSE)</code>:</p>

<pre><code class="r">cleaned_data %&gt;%
  filter(BP == &quot;leucine biosynthesis&quot;) %&gt;%
  ggplot(aes(rate, expression, color = nutrient)) +
  geom_point() +
  geom_smooth(method = &quot;lm&quot;, se = FALSE) +
  facet_wrap(~name)
</code></pre>

<p><img src="assets/fig/unnamed-chunk-6-1.png" alt="plot of chunk unnamed-chunk-6"> </p>

<p>This is great, but there are so many different genes and processes - how can we look at them all?  How do you deal with collaborators who are constantly harassing you asking for plots of different genes?  Shiny is the answer!</p>

<h2>Instructions</h2>

<ul>
<li>Create a new shiny app in RStudio by going File &gt; New Project &gt; New Directory &gt; Shiny Web Application &gt; Give your project a name and hit create (open in a new window might be a good idea too).</li>
<li>Create a new file in the shiny app directory called <code>global.R</code></li>
<li>Paste the following data processing code into <code>global.R</code></li>
</ul>

<pre><code class="r">### Data preparation
require(readr)
require(dplyr)
require(tidyr)
require(ggplot2)
original_data = read_delim(&quot;http://www.maths.usyd.edu.au/u/gartht/Brauer2008_DataSet1.tds&quot;, 
    delim = &quot;\t&quot;)
nutrient_names &lt;- c(G = &quot;Glucose&quot;, L = &quot;Leucine&quot;, P = &quot;Phosphate&quot;, S = &quot;Sulfate&quot;, 
    N = &quot;Ammonia&quot;, U = &quot;Uracil&quot;)
cleaned_data = original_data %&gt;% separate(NAME, c(&quot;name&quot;, &quot;BP&quot;, &quot;MF&quot;, &quot;systematic_name&quot;, 
    &quot;number&quot;), sep = &quot;\\|\\|&quot;) %&gt;% mutate_each(funs(trimws), name:systematic_name) %&gt;% 
    select(-number, -GID, -YORF, -GWEIGHT) %&gt;% gather(sample, expression, G0.05:U0.3) %&gt;% 
    separate(sample, c(&quot;nutrient&quot;, &quot;rate&quot;), sep = 1, convert = TRUE) %&gt;% mutate(nutrient = plyr::revalue(nutrient, 
    nutrient_names)) %&gt;% filter(!is.na(expression), systematic_name != &quot;&quot;)
</code></pre>

<ul>
<li>Whatever&#39;s in <code>global.R</code> will be read and executed before the shiny app loads.  If you want you can download the data set and include it in the same folder as <code>ui.R</code>, <code>server.R</code> and <code>global.R</code> then simplify the <code>read_delim()</code> function to refer to just <code>Brauer2008_DataSet1.tds</code> (without loading it over the internet every time).</li>
<li>Run the app to make sure it&#39;s working in this basic form.</li>
<li><p>Replace the default plot with the first plot we generated above showing results for the <strong>LEU1</strong> gene. Make sure this is working by running the app before proceeding any further.</p></li>
<li><p>You want this in the <code>server.R</code> file</p></li>
</ul>

<pre><code class="r">output$plot1 = renderPlot({
    cleaned_data %&gt;% filter(name == &quot;LEU1&quot;) %&gt;% ggplot(aes(rate, expression, 
        color = nutrient)) + geom_line() + theme_bw(base_size = 14) + facet_wrap(~name + 
        systematic_name)
})
</code></pre>

<ul>
<li>And this in the mainPanel part of the <code>ui.R</code> function:</li>
</ul>

<pre><code class="r">plotOutput(&quot;plot1&quot;)
</code></pre>

<ul>
<li>We want to be able to explore different genes instead of just having the <strong>LEU1</strong> gene plotted.  To do this we need a drop down menu and then we need to feed the results of that drop down menu into the plot.  Add this code chunk to the sidebarPanel section in the <code>ui.R</code> file:</li>
</ul>

<pre><code class="r">selectizeInput(inputId = &quot;gene&quot;, label = &quot;Select gene&quot;, choices = sort(unique(cleaned_data$name)), 
    selected = &quot;LEU1&quot;, multiple = FALSE)
</code></pre>

<ul>
<li>Run to app again to see if it works.  You should get a menu showing up on the left hand side listing all the genes.  This is what the <code>selectizeInput</code> function does - it takes a bunch of <code>choices</code> and offers them to the user as a dropdown list if <code>multiple=FALSE</code> (or a bit fancier where you can select multiple values <code>multiple=TRUE</code>).</li>
<li>Now we need to link the selected gene to the plot.  The selected gene will be accessible though <code>input$gene</code>.  To explain this, <code>input</code> is common to all shiny pachages, it is a named list that contains all the inputs that you define in shiny.  We specified the name <code>gene</code> using the <code>inputId</code> argument to the <code>selectizeInput</code> function.  In the <code>server.R</code> function update the plot function to the following:</li>
</ul>

<pre><code class="r">output$plot1 = renderPlot({
    filter(is.element(name, input$gene)) %&gt;% ggplot(aes(rate, expression, color = nutrient)) + 
        geom_line() + theme_bw(base_size = 14) + facet_wrap(~name + systematic_name)
})
</code></pre>

<ul>
<li><p>What if we wanted to plot multiple genes?  Try changing <code>multiple=TRUE</code> in the selectize input and seeing what happens.  It probably won&#39;t work when multiple genes are selected.  We need to update out filtering in the <code>server.R</code> function to account for multiple gene options.  Something like <code>filter(is.element(name, input$gene))</code> should do the trick.</p></li>
<li><p>What if we wanted to have the option of a line of best fit?  Add in a checkbox to the sidebarPanel so that your sidebarPanel function should now look like this:</p></li>
</ul>

<pre><code class="r">sidebarPanel(selectizeInput(inputId = &quot;gene&quot;, label = &quot;Select gene(s)&quot;, choices = sort(unique(cleaned_data$name)), 
    selected = &quot;LEU1&quot;, multiple = TRUE), checkboxInput(inputId = &quot;line&quot;, label = &quot;Add line of best fit?&quot;, 
    value = FALSE))
</code></pre>

<ul>
<li>Think about how to incorporate this logical input into the <code>server.R</code> code.  When the checkbox is ticked <code>input$line</code> will return a value of <code>TRUE</code>.  When unticked, <code>input$line</code> returns a value of <code>FALSE</code>.  It makes sense to use an <code>if</code> statement to get two different behaviours.  In the <code>server.R</code> adapt your code to something like:</li>
</ul>

<pre><code class="r">output$plot1 = renderPlot({
    if (input$line) {
        cleaned_data %&gt;% filter(is.element(name, input$gene)) %&gt;% ggplot(aes(rate, 
            expression, color = nutrient)) + geom_point() + theme_bw(base_size = 14) + 
            geom_smooth(method = &quot;lm&quot;, se = FALSE) + facet_wrap(~name)
    } else {
        cleaned_data %&gt;% filter(is.element(name, input$gene)) %&gt;% ggplot(aes(rate, 
            expression, color = nutrient)) + geom_line() + theme_bw(base_size = 14) + 
            facet_wrap(~name + systematic_name)
    }
})
</code></pre>

<ul>
<li>OK so at this point, you should be able to plot multiple genes and toggle between the ploting just the raw data and fitting simple linear regression lines (with the raw observations in the background).</li>
<li>What if we wanted to be able to select all the genes related to a particular biological process.  We&#39;ll need a new dropdown menu and a new plot. Do this by adapting the above code.  Optionally you can include it in a new tab using <code>tabsetPanel</code> in the <code>ui.R</code> file - you can see how this is done in the solutions, but this is probably more advanced than we really need for now.</li>
<li>Once you&#39;re happy with your app, head over to <a href="http://www.shinyapps.io/">shinyapps.io</a>, create an account and follow the instructions to push your app to the web so you can share the link and show off your fine work to all your collaborators!  To do this, you&#39;ll first need to install these packages:</li>
</ul>

<pre><code class="r">install.packages(c(&quot;PKI&quot;, &quot;packrat&quot;, &quot;rsconnect&quot;))
</code></pre>

<h3>Suggested solution</h3>

<p>See <a href="https://github.com/garthtarr/visR/tree/gh-pages/labs/01/myapp">here</a>. Try not to cheat, unless you really have to.</p>

<h2>Reference</h2>

<ul>
<li>Robinson D (2015). &quot;Cleaning and visualizing genomic data: a case study in tidy analysis&quot;, blog post. <a href="http://varianceexplained.org/r/tidy-genomics/">http://varianceexplained.org/r/tidy-genomics/</a></li>
<li>Brauer et al. (2008). Coordination of Growth Rate, Cell Cycle, Stress Response, and Metabolic Activity in Yeast,</li>
</ul>

      </section>
      <footer>
        <p>Maintained by <a href="http://github.com/garthtarr"><i class="fa fa-github"></i>/garthtarr
        </a></p>
        <!--<p><small>Hosted on GitHub Pages &mdash; Theme by 
          <a href="https://github.com/orderedlist">orderedlist</a></small>
        </p>-->
      </footer>    </div>
    <script src="../../libraries/frameworks/minimal/javascripts/scale.fix.js"></script>
  </body>
  <!-- Load Javascripts for Widgets -->
  
  <!-- MathJax: Fall back to local if CDN offline but local image fonts are not supported (saves >100MB) -->
  <script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      tex2jax: {
        inlineMath: [['$','$'], ['\\(','\\)']],
        processEscapes: true
      }
    });
  </script>
  <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/2.0-latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
  <!-- <script src="https://c328740.ssl.cf1.rackcdn.com/mathjax/2.0-latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
  </script> -->
  <script>window.MathJax || document.write('<script type="text/x-mathjax-config">MathJax.Hub.Config({"HTML-CSS":{imageFont:null}});<\/script><script src="../../libraries/widgets/mathjax/MathJax.js?config=TeX-AMS-MML_HTMLorMML"><\/script>')
</script>
<!-- LOAD HIGHLIGHTER JS FILES -->
  <script src="../../libraries/highlighters/highlight.js/highlight.pack.js"></script>
  <script>hljs.initHighlightingOnLoad();</script>
  <!-- DONE LOADING HIGHLIGHTER JS FILES -->
   
  </html>